File: backend/internal/service/transform_test.go
package service

import (
	"math"
	"testing"
	"time"

	"github.com/naotama2002/habity/backend/internal/habitify"
)

func TestTransformHabit_BasicCheck(t *testing.T) {
	h := habitify.Habit{
		ID:         "h-1",
		Name:       "Morning Run",
		IsArchived: false,
		StartDate:  time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC),
		TimeOfDay:  []string{"morning"},
		LogMethod:  "check",
		Priority:   2,
		Recurrence: "RRULE:FREQ=DAILY",
		Goal:       &habitify.Goal{UnitType: "count", Value: 1, Periodicity: "daily"},
		Area:       &habitify.Area{ID: "area-1", Name: "Health"},
	}

	categoryMap := map[string]string{"area-1": "cat-uuid-1"}
	result := TransformHabit(h, "user-1", categoryMap)

	if result.UserID != "user-1" {
		t.Errorf("expected user-1, got %s", result.UserID)
	}
	if result.Name != "Morning Run" {
		t.Errorf("expected Morning Run, got %s", result.Name)
	}
	if result.TrackingType != "boolean" {
		t.Errorf("expected boolean, got %s", result.TrackingType)
	}
	if result.GoalValue != 1 {
		t.Errorf("expected goal value 1, got %f", result.GoalValue)
	}
	if result.GoalUnit != "times" {
		t.Errorf("expected times, got %s", result.GoalUnit)
	}
	if result.GoalPeriod != "daily" {
		t.Errorf("expected daily, got %s", result.GoalPeriod)
	}
	if result.Status != "active" {
		t.Errorf("expected active, got %s", result.Status)
	}
	if result.StartDate != "2024-01-01" {
		t.Errorf("expected 2024-01-01, got %s", result.StartDate)
	}
	if result.ExternalID != "h-1" {
		t.Errorf("expected h-1, got %s", result.ExternalID)
	}
	if result.ExternalSource != "habitify" {
		t.Errorf("expected habitify, got %s", result.ExternalSource)
	}
	if result.CategoryID == nil || *result.CategoryID != "cat-uuid-1" {
		t.Error("expected category_id to be cat-uuid-1")
	}
	if result.RecurrenceRule == nil || *result.RecurrenceRule != "RRULE:FREQ=DAILY" {
		t.Error("expected recurrence rule")
	}
	if len(result.TimeOfDay) != 1 || result.TimeOfDay[0] != "morning" {
		t.Errorf("expected [morning], got %v", result.TimeOfDay)
	}
	if result.SortOrder != 2 {
		t.Errorf("expected sort_order 2, got %d", result.SortOrder)
	}
}

func TestTransformHabit_Archived(t *testing.T) {
	h := habitify.Habit{
		ID:         "h-2",
		Name:       "Old Habit",
		IsArchived: true,
		StartDate:  time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC),
		LogMethod:  "check",
	}

	result := TransformHabit(h, "user-1", nil)
	if result.Status != "archived" {
		t.Errorf("expected archived, got %s", result.Status)
	}
}

func TestTransformHabit_LogMethodMapping(t *testing.T) {
	tests := []struct {
		logMethod    string
		expectedType string
	}{
		{"check", "boolean"},
		{"measure", "numeric"},
		{"number", "numeric"},
		{"timer", "duration"},
		{"unknown", "boolean"},
	}

	for _, tt := range tests {
		t.Run(tt.logMethod, func(t *testing.T) {
			h := habitify.Habit{
				ID:        "h-1",
				Name:      "Test",
				StartDate: time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC),
				LogMethod: tt.logMethod,
			}
			result := TransformHabit(h, "user-1", nil)
			if result.TrackingType != tt.expectedType {
				t.Errorf("log_method %s: expected %s, got %s", tt.logMethod, tt.expectedType, result.TrackingType)
			}
		})
	}
}

func TestTransformHabit_GoalMapping(t *testing.T) {
	tests := []struct {
		name       string
		goal       *habitify.Goal
		wantValue  float64
		wantUnit   string
		wantPeriod string
	}{
		{
			name:       "nil goal defaults",
			goal:       nil,
			wantValue:  1,
			wantUnit:   "times",
			wantPeriod: "daily",
		},
		{
			name:       "count daily",
			goal:       &habitify.Goal{UnitType: "count", Value: 5, Periodicity: "daily"},
			wantValue:  5,
			wantUnit:   "times",
			wantPeriod: "daily",
		},
		{
			name:       "minute weekly",
			goal:       &habitify.Goal{UnitType: "minute", Value: 30, Periodicity: "weekly"},
			wantValue:  30,
			wantUnit:   "min",
			wantPeriod: "weekly",
		},
		{
			name:       "kilometer monthly",
			goal:       &habitify.Goal{UnitType: "kilometer", Value: 100, Periodicity: "monthly"},
			wantValue:  100,
			wantUnit:   "km",
			wantPeriod: "monthly",
		},
		{
			name:       "unknown unit type passes through",
			goal:       &habitify.Goal{UnitType: "custom_unit", Value: 10, Periodicity: "daily"},
			wantValue:  10,
			wantUnit:   "custom_unit",
			wantPeriod: "daily",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			h := habitify.Habit{
				ID:        "h-1",
				Name:      "Test",
				StartDate: time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC),
				LogMethod: "check",
				Goal:      tt.goal,
			}
			result := TransformHabit(h, "user-1", nil)
			if result.GoalValue != tt.wantValue {
				t.Errorf("expected value %f, got %f", tt.wantValue, result.GoalValue)
			}
			if result.GoalUnit != tt.wantUnit {
				t.Errorf("expected unit %s, got %s", tt.wantUnit, result.GoalUnit)
			}
			if result.GoalPeriod != tt.wantPeriod {
				t.Errorf("expected period %s, got %s", tt.wantPeriod, result.GoalPeriod)
			}
		})
	}
}

func TestTransformHabit_NullArea(t *testing.T) {
	h := habitify.Habit{
		ID:        "h-1",
		Name:      "Test",
		StartDate: time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC),
		LogMethod: "check",
		Area:      nil,
	}

	result := TransformHabit(h, "user-1", nil)
	if result.CategoryID != nil {
		t.Error("expected nil category_id for nil area")
	}
}

func TestTransformHabit_EmptyTimeOfDay(t *testing.T) {
	h := habitify.Habit{
		ID:        "h-1",
		Name:      "Test",
		StartDate: time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC),
		LogMethod: "check",
		TimeOfDay: []string{},
	}

	result := TransformHabit(h, "user-1", nil)
	if len(result.TimeOfDay) != 1 || result.TimeOfDay[0] != "anytime" {
		t.Errorf("expected [anytime] for empty time_of_day, got %v", result.TimeOfDay)
	}
}

func TestTransformHabit_InvalidTimeOfDayFiltered(t *testing.T) {
	h := habitify.Habit{
		ID:        "h-1",
		Name:      "Test",
		StartDate: time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC),
		LogMethod: "check",
		TimeOfDay: []string{"morning", "invalid_value", "evening"},
	}

	result := TransformHabit(h, "user-1", nil)
	if len(result.TimeOfDay) != 2 {
		t.Fatalf("expected 2 items, got %v", result.TimeOfDay)
	}
	if result.TimeOfDay[0] != "morning" || result.TimeOfDay[1] != "evening" {
		t.Errorf("expected [morning, evening], got %v", result.TimeOfDay)
	}
}

func TestTransformHabit_NoRecurrence(t *testing.T) {
	h := habitify.Habit{
		ID:        "h-1",
		Name:      "Test",
		StartDate: time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC),
		LogMethod: "check",
	}

	result := TransformHabit(h, "user-1", nil)
	if result.RecurrenceRule != nil {
		t.Error("expected nil recurrence rule for empty string")
	}
}

func TestTransformHabit_OverflowPriority(t *testing.T) {
	tests := []struct {
		name     string
		priority float64
		want     int
	}{
		{"normal", 5, 5},
		{"zero", 0, 0},
		{"negative normal", -1, -1},
		{"very large", 1e18, 0},
		{"very small", -1e18, 0},
		{"NaN", math.NaN(), 0},
		{"positive Inf", math.Inf(1), 0},
		{"negative Inf", math.Inf(-1), 0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			h := habitify.Habit{
				ID:        "h-1",
				Name:      "Test",
				StartDate: time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC),
				LogMethod: "check",
				Priority:  tt.priority,
			}
			result := TransformHabit(h, "user-1", nil)
			if result.SortOrder != tt.want {
				t.Errorf("priority %v: expected sort_order %d, got %d", tt.priority, tt.want, result.SortOrder)
			}
		})
	}
}

func TestTransformLog(t *testing.T) {
	l := habitify.Log{
		ID:          "log-1",
		HabitID:     "h-1",
		Value:       1,
		CreatedDate: time.Date(2024, 1, 15, 10, 30, 0, 0, time.UTC),
		UnitType:    "count",
	}

	result := TransformLog(l, "habity-habit-1", "user-1", time.UTC)

	if result.UserID != "user-1" {
		t.Errorf("expected user-1, got %s", result.UserID)
	}
	if result.HabitID != "habity-habit-1" {
		t.Errorf("expected habity-habit-1, got %s", result.HabitID)
	}
	if result.Value != 1 {
		t.Errorf("expected value 1, got %f", result.Value)
	}
	if result.TargetDate != "2024-01-15" {
		t.Errorf("expected 2024-01-15, got %s", result.TargetDate)
	}
	if result.Status != "completed" {
		t.Errorf("expected completed, got %s", result.Status)
	}
	if result.ExternalID != "log-1" {
		t.Errorf("expected log-1, got %s", result.ExternalID)
	}
	if !result.CompletedAt.Equal(l.CreatedDate) {
		t.Errorf("expected completed_at to equal created_date")
	}
}

func TestTransformLog_TimezoneConversion(t *testing.T) {
	jst := time.FixedZone("JST", 9*60*60)

	tests := []struct {
		name       string
		utcTime    time.Time
		loc        *time.Location
		wantDate   string
	}{
		{
			name:     "UTC midnight stays same date in UTC",
			utcTime:  time.Date(2026, 2, 10, 0, 0, 0, 0, time.UTC),
			loc:      time.UTC,
			wantDate: "2026-02-10",
		},
		{
			name:     "UTC 23:00 becomes next day in JST",
			utcTime:  time.Date(2026, 2, 9, 23, 0, 0, 0, time.UTC),
			loc:      jst,
			wantDate: "2026-02-10",
		},
		{
			name:     "UTC 14:00 stays same day in JST (23:00 JST)",
			utcTime:  time.Date(2026, 2, 10, 14, 0, 0, 0, time.UTC),
			loc:      jst,
			wantDate: "2026-02-10",
		},
		{
			name:     "UTC 15:00 becomes next day in JST (00:00 JST)",
			utcTime:  time.Date(2026, 2, 10, 15, 0, 0, 0, time.UTC),
			loc:      jst,
			wantDate: "2026-02-11",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			l := habitify.Log{
				ID:          "log-1",
				HabitID:     "h-1",
				Value:       1,
				CreatedDate: tt.utcTime,
			}
			result := TransformLog(l, "habity-habit-1", "user-1", tt.loc)
			if result.TargetDate != tt.wantDate {
				t.Errorf("expected target_date %s, got %s", tt.wantDate, result.TargetDate)
			}
		})
	}
}

File: src/lib/habitify/__tests__/transform.test.ts
import { describe, expect, it } from '@jest/globals';

import {
  transformHabit,
  transformDailyProgress,
  findActiveGoal,
  inferTrackingType,
  mapGoal,
  mapGoalUnit,
  mapGoalPeriod,
  occurrenceToRRule,
  mapTimeOfDays,
} from '../transform';
import type { HabitifyHabit, HabitifyGoal, HabitifyDailyProgress } from '../types';

// ─── Helper: minimal v2 habit ─────────────────────────

function makeHabit(overrides: Partial<HabitifyHabit> = {}): HabitifyHabit {
  return {
    id: 'h-1',
    name: 'Morning Run',
    occurrence: { type: 'daily' },
    startDate: '2024-01-01',
    createdAt: '2024-01-01T00:00:00Z',
    isArchived: false,
    logMethod: 'manual',
    goals: [
      {
        id: 'goal-1',
        createdAt: '2024-01-01T00:00:00Z',
        periodicity: 'daily',
        value: 1,
        unit: 'rep',
        isActive: true,
      },
    ],
    areas: [{ id: 'area-1', name: 'Health' }],
    timeOfDays: [{ id: 'tod-1', name: 'Morning' }],
    ...overrides,
  };
}

function makeGoal(overrides: Partial<HabitifyGoal> = {}): HabitifyGoal {
  return {
    id: 'goal-1',
    createdAt: '2024-01-01T00:00:00Z',
    periodicity: 'daily',
    value: 1,
    unit: 'rep',
    isActive: true,
    ...overrides,
  };
}

// ─── transformHabit ───────────────────────────────────

describe('transformHabit', () => {
  it('should transform a basic v2 habit', () => {
    const h = makeHabit();
    const categoryMap = { 'area-1': 'cat-uuid-1' };
    const result = transformHabit(h, 'user-1', categoryMap);

    expect(result.user_id).toBe('user-1');
    expect(result.name).toBe('Morning Run');
    // Default goal is 1 rep — Habitify's checkbox-style habit shape.
    expect(result.tracking_type).toBe('boolean');
    expect(result.goal_value).toBe(1);
    expect(result.goal_unit).toBe('times');
    expect(result.goal_period).toBe('daily');
    expect(result.status).toBe('active');
    expect(result.start_date).toBe('2024-01-01');
    expect(result.external_id).toBe('h-1');
    expect(result.external_source).toBe('habitify');
    expect(result.category_id).toBe('cat-uuid-1');
    expect(result.recurrence_rule).toBe('RRULE:FREQ=DAILY');
    expect(result.time_of_day).toEqual(['morning']);
    expect(result.end_date).toBeNull();
  });

  it('should set status to archived when isArchived is true', () => {
    const result = transformHabit(makeHabit({ isArchived: true }), 'user-1', {});
    expect(result.status).toBe('archived');
  });

  it('should set category_id to null when areas is empty', () => {
    const result = transformHabit(makeHabit({ areas: [] }), 'user-1', {});
    expect(result.category_id).toBeNull();
  });

  it('should use first area for category mapping', () => {
    const h = makeHabit({
      areas: [
        { id: 'area-1', name: 'Health' },
        { id: 'area-2', name: 'Work' },
      ],
    });
    const categoryMap = { 'area-1': 'cat-1', 'area-2': 'cat-2' };
    const result = transformHabit(h, 'user-1', categoryMap);
    expect(result.category_id).toBe('cat-1');
  });

  it('should default time_of_day to ["anytime"] when timeOfDays is empty', () => {
    const result = transformHabit(makeHabit({ timeOfDays: [] }), 'user-1', {});
    expect(result.time_of_day).toEqual(['anytime']);
  });

  it('should use boolean tracking_type when no goals', () => {
    const result = transformHabit(makeHabit({ goals: [] }), 'user-1', {});
    expect(result.tracking_type).toBe('boolean');
    expect(result.goal_value).toBe(1);
    expect(result.goal_unit).toBe('times');
  });

  it('should use duration tracking_type for time-based goals', () => {
    const h = makeHabit({
      goals: [makeGoal({ unit: 'min', value: 30 })],
    });
    const result = transformHabit(h, 'user-1', {});
    expect(result.tracking_type).toBe('duration');
    expect(result.goal_unit).toBe('min');
    expect(result.goal_value).toBe(30);
  });

  it('should use numeric tracking_type for a rep goal with value > 1', () => {
    const h = makeHabit({
      goals: [makeGoal({ unit: 'rep', value: 5 })],
    });
    const result = transformHabit(h, 'user-1', {});
    expect(result.tracking_type).toBe('numeric');
    expect(result.goal_value).toBe(5);
  });

  it('should convert weekDays occurrence to RRULE', () => {
    const h = makeHabit({
      occurrence: { type: 'weekDays', days: [1, 3, 5] },
    });
    const result = transformHabit(h, 'user-1', {});
    expect(result.recurrence_rule).toBe('RRULE:FREQ=WEEKLY;BYDAY=MO,WE,FR');
  });

  it('should convert intervalDays occurrence to RRULE', () => {
    const h = makeHabit({
      occurrence: { type: 'intervalDays', interval: 3 },
    });
    const result = transformHabit(h, 'user-1', {});
    expect(result.recurrence_rule).toBe('RRULE:FREQ=DAILY;INTERVAL=3');
  });

  it('should use description from v2 habit', () => {
    const h = makeHabit({ description: 'Run every morning' });
    const result = transformHabit(h, 'user-1', {});
    expect(result.description).toBe('Run every morning');
  });

  it('should import the habit with null goal fields when periodicity is unmappable', () => {
    const h = makeHabit({
      goals: [makeGoal({ periodicity: 'yearly', value: 5, unit: 'kM' })],
    });
    const result = transformHabit(h, 'user-1', {});
    expect(result.goal_value).toBeNull();
    expect(result.goal_unit).toBeNull();
    expect(result.goal_period).toBeNull();
    // The habit itself is still imported.
    expect(result.name).toBe('Morning Run');
    expect(result.external_id).toBe('h-1');
  });
});

// ─── transformDailyProgress ───────────────────────────

describe('transformDailyProgress', () => {
  it('should transform a completed daily progress', () => {
    const dp: HabitifyDailyProgress = {
      date: '2024-01-15',
      totalLog: 3.5,
      status: 'completed',
    };

    const result = transformDailyProgress(dp, 'habity-h-1', 'user-1');

    expect(result).not.toBeNull();
    expect(result!.user_id).toBe('user-1');
    expect(result!.habit_id).toBe('habity-h-1');
    expect(result!.value).toBe(3.5);
    expect(result!.target_date).toBe('2024-01-15');
    expect(result!.completed_at).toBe('2024-01-15T00:00:00');
    expect(result!.status).toBe('completed');
    expect(result!.external_id).toBeNull();
  });

  it('should transform a skipped daily progress', () => {
    const dp: HabitifyDailyProgress = {
      date: '2024-01-16',
      totalLog: 0,
      status: 'skipped',
    };

    const result = transformDailyProgress(dp, 'habity-h-1', 'user-1');

    expect(result).not.toBeNull();
    expect(result!.status).toBe('skipped');
    expect(result!.value).toBe(0);
  });

  it('should return null for failed status', () => {
    const dp: HabitifyDailyProgress = {
      date: '2024-01-17',
      totalLog: 0,
      status: 'failed',
    };

    expect(transformDailyProgress(dp, 'habity-h-1', 'user-1')).toBeNull();
  });

  it('should return null for inprogress status', () => {
    const dp: HabitifyDailyProgress = {
      date: '2024-01-18',
      totalLog: 2,
      status: 'inprogress',
    };

    expect(transformDailyProgress(dp, 'habity-h-1', 'user-1')).toBeNull();
  });
});

// ─── findActiveGoal ───────────────────────────────────

describe('findActiveGoal', () => {
  it('should return the active goal', () => {
    const goals = [
      makeGoal({ id: 'g1', isActive: false }),
      makeGoal({ id: 'g2', isActive: true }),
    ];
    expect(findActiveGoal(goals)!.id).toBe('g2');
  });

  it('should fall back to first goal when none active', () => {
    const goals = [
      makeGoal({ id: 'g1', isActive: false }),
      makeGoal({ id: 'g2', isActive: false }),
    ];
    expect(findActiveGoal(goals)!.id).toBe('g1');
  });

  it('should return null for empty goals', () => {
    expect(findActiveGoal([])).toBeNull();
  });
});

// ─── inferTrackingType ────────────────────────────────

describe('inferTrackingType', () => {
  it.each([
    // Duration units always map to 'duration', regardless of goal value.
    { unit: 'min', value: 1, expected: 'duration' },
    { unit: 'min', value: 30, expected: 'duration' },
    { unit: 'hr', value: 1, expected: 'duration' },
    { unit: 'sec', value: 1, expected: 'duration' },
    { unit: 'ms', value: 1, expected: 'duration' },
    // 'rep' + value 1 is Habitify's checkbox ("just do it") shape → boolean.
    { unit: 'rep', value: 1, expected: 'boolean' },
    // 'rep' with any other value is a real numeric goal (e.g. "5 reps").
    { unit: 'rep', value: 2, expected: 'numeric' },
    { unit: 'rep', value: 5, expected: 'numeric' },
    // Non-'rep' units are never treated as boolean, even at value 1.
    { unit: 'step', value: 1, expected: 'numeric' },
    { unit: 'step', value: 100, expected: 'numeric' },
    { unit: 'kM', value: 1, expected: 'numeric' },
    { unit: 'm', value: 1, expected: 'numeric' },
  ])(
    'should infer "$expected" for unit "$unit" and value $value',
    ({ unit, value, expected }) => {
      expect(inferTrackingType(makeGoal({ unit, value }))).toBe(expected);
    },
  );

  it('should return boolean when goal is null', () => {
    expect(inferTrackingType(null)).toBe('boolean');
  });
});

// ─── mapGoal ──────────────────────────────────────────

describe('mapGoal', () => {
  it('should return defaults for null goal', () => {
    const result = mapGoal(null);
    expect(result).toEqual({ value: 1, unit: 'times', period: 'daily' });
  });

  it('should map a v2 goal', () => {
    const result = mapGoal(makeGoal({ value: 5, unit: 'kM', periodicity: 'weekly' }));
    expect(result).toEqual({ value: 5, unit: 'km', period: 'weekly' });
  });

  it('should not adopt the goal (null value/unit/period) when periodicity is unknown', () => {
    const result = mapGoal(makeGoal({ value: 5, unit: 'kM', periodicity: 'yearly' }));
    expect(result).toEqual({ value: null, unit: null, period: null });
  });
});

// ─── mapGoalUnit ──────────────────────────────────────

describe('mapGoalUnit', () => {
  it.each([
    ['rep', 'times'],
    ['min', 'min'],
    ['hr', 'hours'],
    ['sec', 'sec'],
    ['kM', 'km'],
    ['m', 'm'],
    ['step', 'steps'],
    ['floor', 'floors'],
    ['kCal', 'kcal'],
    ['L', 'L'],
    ['mL', 'mL'],
    ['kg', 'kg'],
    ['unknown_unit', 'unknown_unit'],
  ])('should map "%s" to "%s"', (input, expected) => {
    expect(mapGoalUnit(input)).toBe(expected);
  });
});

// ─── mapGoalPeriod ────────────────────────────────────

describe('mapGoalPeriod', () => {
  it.each([
    ['daily', 'daily'],
    ['weekly', 'weekly'],
    ['monthly', 'monthly'],
  ])('should map "%s" to "%s"', (input, expected) => {
    expect(mapGoalPeriod(input)).toBe(expected);
  });

  it.each(['yearly', 'unknown'])(
    'should return null for unmappable periodicity "%s" instead of fabricating "daily"',
    (input) => {
      expect(mapGoalPeriod(input)).toBeNull();
    },
  );
});

// ─── occurrenceToRRule ────────────────────────────────

describe('occurrenceToRRule', () => {
  it('should convert daily to RRULE:FREQ=DAILY', () => {
    expect(occurrenceToRRule({ type: 'daily' })).toBe('RRULE:FREQ=DAILY');
  });

  it('should convert weekDays to RRULE with BYDAY', () => {
    expect(occurrenceToRRule({ type: 'weekDays', days: [0, 1, 6] })).toBe(
      'RRULE:FREQ=WEEKLY;BYDAY=SU,MO,SA',
    );
  });

  it('should convert weekDays [1,3,5] to MO,WE,FR', () => {
    expect(occurrenceToRRule({ type: 'weekDays', days: [1, 3, 5] })).toBe(
      'RRULE:FREQ=WEEKLY;BYDAY=MO,WE,FR',
    );
  });

  it('should convert intervalDays to RRULE with INTERVAL', () => {
    expect(occurrenceToRRule({ type: 'intervalDays', interval: 3 })).toBe(
      'RRULE:FREQ=DAILY;INTERVAL=3',
    );
  });

  it('should handle all 7 days', () => {
    expect(
      occurrenceToRRule({ type: 'weekDays', days: [0, 1, 2, 3, 4, 5, 6] }),
    ).toBe('RRULE:FREQ=WEEKLY;BYDAY=SU,MO,TU,WE,TH,FR,SA');
  });

  it('should fall back to RRULE:FREQ=DAILY when weekDays.days has only out-of-range values', () => {
    expect(occurrenceToRRule({ type: 'weekDays', days: [7, -1] })).toBe(
      'RRULE:FREQ=DAILY',
    );
  });

  it('should return null for an unknown occurrence type', () => {
    expect(occurrenceToRRule({ type: 'timesPerWeek', times: 3 })).toBeNull();
  });
});

// ─── mapTimeOfDays ────────────────────────────────────

describe('mapTimeOfDays', () => {
  it('should return ["anytime"] for empty array', () => {
    expect(mapTimeOfDays([])).toEqual(['anytime']);
  });

  it('should map time-of-day names to lowercase', () => {
    expect(
      mapTimeOfDays([
        { id: '1', name: 'Morning' },
        { id: '2', name: 'Evening' },
      ]),
    ).toEqual(['morning', 'evening']);
  });

  it('should filter invalid values', () => {
    expect(
      mapTimeOfDays([
        { id: '1', name: 'Morning' },
        { id: '2', name: 'CustomTime' },
        { id: '3', name: 'Night' },
      ]),
    ).toEqual(['morning', 'night']);
  });

  it('should return ["anytime"] when all values are invalid', () => {
    expect(mapTimeOfDays([{ id: '1', name: 'CustomTime' }])).toEqual([
      'anytime',
    ]);
  });
});
